design-proposals/klip-24-key-column-semantics-in-queries.md

```diff
@@ -0,0 +1,372 @@
+```
+
+### Any key name
+
+Persistent queries failing if the projection contained the key column existed before the work to
+[allow any key name](https://github.com/confluentinc/ksql/issues/3536), though this work also
+exacerbated the situation, as the key column is no longer always called `ROWKEY`. Instead, users can
+pick any name for the key column, and the key column takes on the name of any `PARTITION BY`,
+`GROUP BY` or `JOIN ON` on a single column. This causes seemingly correct and common patterns to
+fail with duplicate column errors:
+
+```sql
+-- Before 'any key name':
+CREATE STREAM INPUT (ROWKEY INT KEY, V0 INT, V1 INT) WITH (...);
+CREATE TABLE OUTPUT AS SELECT V0, COUNT(*) AS COUNT FROM INPUT GROUP BY V0;
+-- resulting schema: ROwKEY INT KEY, V0 INT, COUNT BIGINT
+-- Note: ROWKEY & V0 store the same data, which isn't ideal.
+
+-- vs --
+
+-- After 'any key name':
+CREATE STREAM INPUT (ID INT KEY, V0 INT, V1 INT) WITH (...);
+CREATE TABLE OUTPUT AS SELECT V0, COUNT(*) AS COUNT FROM INPUT GROUP BY V0;
+-- fails with duplicate column error on 'V0'.
+
+-- Note: the equivalent query would of failed before the any-key-name work:
+CREATE TABLE OUTPUT AS SELECT ROWKEY, COUNT(*) AS COUNT FROM INPUT GROUP BY V0;
+-- fails with duplicate column error on 'ROWKEY'.
+```
+
+As you can see from above, the common pattern of selecting the group by key and the aggregate fails.
+To 'fix' the query the user must remove `V0` from the projection - which is counter intuitive, as
+the user wishes this column in the result.
+
+We propose that this query should work, without any modification, and without storing duplicate
+data.
+
+### Requiring the key column in a projection in persistent queries
+
+By flipping the semantics so that the key column is not _implicitly_ copied across in persistent
+queries we place the burden on the user to _explicitly_ specify it in the projection.
+
+Where as it is possible to have a transient query without the key column in the projection, as a
+transient query is simply returning tabular data and has no concept of a key column, persistent
+queries do have the concept of a key column, and changing that column is only allowed via an
+explicit `PARTITION BY` or `GROUP BY` statement.  Therefore users will _always_ have to specify the
+key column in the projection of persistent queries.
+
+For example, the following statements, which previously executed, will now fail:
+
+```sql
+CREATE TABLE OUTPUT AS SELECT V0, V1 FROM INPUT;
+-- fails as key column not in projection.
+
+CREATE TABLE OUTPUT AS SELECT COUNT(*) FROM INPUT GROUP BY V1;
+-- fails as key column not in projection.
+
+CREATE STREAM OUTPUT AS SELECT V0, ID FROM INPUT PARTITION BY V1;
+-- fails as key column not in projection.
+
+CREATE TABLE OUTPUT AS SELECT I1.V0, I2.V2 FROM INPUT I1 JOIN INPUT I2 ON I1.ID = I2.ID;
+-- fails as key column not in projection.
+```
+
+We propose this is acceptable, and preferable to the current model.
```

Comment:
what do you think about having it so that if the user does not include the key field, they simply cannot access it? That way they don't need to explicitly specify it in the project, they just won't be able to access it otherwise. 

I think this is reasonable, as with plain SQL I wouldn't be able to access the GROUP BY field (for example) if I didn't include it in the projection (even if you `CREATE TABLE AS SELECT ... GROUP BY foo`)

```sql
postgres=# CREATE TABLE input (id VARCHAR, count INTEGER);
CREATE TABLE
postgres=# INSERT INTO input (id, count) VALUES ('a', 1);
INSERT 0 1
postgres=# INSERT INTO input (id, count) VALUES ('b', 1);
INSERT 0 1
postgres=# CREATE TABLE agg AS SELECT SUM(count) FROM input GROUP BY id;
SELECT 1
--- notice that the ID column is nowhere to be found, even though it is the "key" for
--- the table below
postgres=# \d agg
                Table "public.agg"
 Column |  Type  | Collation | Nullable | Default
--------+--------+-----------+----------+---------
 sum    | bigint |           |          |
postgres=# SELECT * FROM agg;
 sum
-----
   1
   1
```

Reply:
I'm not sure I follow.

What do you mean by 'can not access it'?

I don't see a similarity between what you're doing in postgres with ksql.  In postgres the table 'agg' suerly doesn't have an `ID` column.  It's just a table containing the result of the query, which returns only a single column: 'sum'.

Reply:
OK, so you're talking about the select, not the create table.

Reply:
So... for others, @agavra proposing that we support something like:

```sql
CREATE TABLE BAR AS SELECT COUNT(*) FROM FOO GROUP BY ID;
```

Which seems totally reasonable and something we should support!

Under the hood we still need to track the `ID` key, but that's just an implementation detail.

```sql
SELECT * FROM BAR;
```

Would return only the count column.  The ID column would not be available in any downstream queries.

This is a good suggestion, but a much bigger change. Probably best left out of scope.

Todo(ac): Consider including in the klip, but not implementing initially.

Reply:
The proposal is quite elegant. I had some initial concerns about downstream ramification (for example, table `BAR` could not be used in a join), but it seems acceptable.

In general, people who are used to RDBMS were often "confused" in the past that we enforce a PK and thus, if we can relax this restriction it might be good.

However,  the proposed change would introduce some "gap" between what people see "within ksqlDB" and what is in the topic: if one uses a consumer to read the table topic, the key would still be there. It might be ok, I just want to raise awareness about it. This would also allow users to "resurrect" a key -- when defining a new table of the topic one can just "re-declare" the key in the DDL statement.

Reply:
Yep, agreed.

I'm torn between including @agavra's suggestion in the KLIP, but initially not implementing it (as its a bigger change), and just covering it in rejected alternatives as possible future work.

Basically, I'd like to avoid too much discussion on this cool / valid suggestion as I don't want it slowing down the approval of this KLIP.

Reply:
Updated to outline this approach as a 'rejected alternative', that's rejected purely to limit scope of change.

Reply:
@rmoff -- the query you describe is supported by any RDBMS I am aware of (correct me if I am wrong)? I totally agree that the query does not make much sense, you just see a bunch of numbers (ie, the counts) but you don't know which count belongs to which ID...

We can argue, that we don't want to allow "less useful" queries like this one at all. But it is quite a "gap" compared to what RDBMS allow users to do. Would be very helpful to learn more from you about it? Is the current limitation of ksqlDB to always require a PK on a table a big deal? If not, maybe we can just keep it this requirement (for good reasons; if there is no PK, a table does not make much sense...)?

Reply:
I didn't realise it's supported by RDBMS. I would argue this both ways:

👉 On one hand, if they support it and it's valid SQL, I guess we should if it's easy to do

👈 On the other hand, as you say, `if there is no PK, a table does not make much sense`, and building this guardrail in would save users from themselves in building an object which they can't use. On an RDBMS I can imagine people building reports with some static SQL to pull back figures and dumping into a front end, i.e. I can see _why_ people might use this shortcut. On a streaming database, it just doesn't make sense to me - and learning from the JOIN / KEY issue that plagued new users to KSQL since its inception, I'd suggest we err on the side of saving users from themselves at this stage.

Reply:
> if there is no PK, a table does not make much sense

Remember, it's a MV really, not a table. It's totally valid to have an MV without a key column. (Internally, the DB might track one for implementation details).

Reply:
IMHO, the argument also holds for a MV. If I just see a "list of numbers" how can I use it in my application? The result of the query only tells you how many unique `ID`s are there (`GROUP BY ID`) (ie, many many output row do I get) and what counts exists. But you don't know which count belongs to which ID. Not sure how this is useful?

Reply:
If we support lookups on things other than primary keys one day, then there are still lots of columns that can add value that aren't the key

Reply:
Not sure if I can follow? Can you elabrate?